storage/mysql/sql.go

```diff
@@ -124,5 +128,30 @@ func readTree(r row) (*trillian.Tree, error) {
 		}
 	}
 
+	// We're going to try to interpret PublicKey as storageSettings, but it could be a
+	// public key from a really old tree, or an empty column from a tree created in the
+	// period between Trillian key material being removed and this column being used for
+	// storing settings.
+	buff := bytes.NewBuffer(publicKey)
+	dec := gob.NewDecoder(buff)
+	ss := &storageSettings{}
+	var o *mysqlpb.StorageOptions
+	if err := dec.Decode(ss); err != nil {
```

Comment:
As the comments above and below here say, there are the options that this column contains one of {public key, nothing, storageSettings}. An errror will occur in either of the first two cases, which maps to having a revisioned tree (which is what we want). The third case should parse properly, and then switch on the storageSettings.Revisioned field (also what we want).

There is another option which is that database corruption mangled a valid storageSettings. At this point it is somewhat normal to expect things to die. But actually I believe this will work well. I've reasoned that going from revisionless to revisioned will work (a mangled storageSettings will either fail to parse, or in an extreme case will be mangled to another valid storageSettings where the revisioned bit is true; either still leads to a working outcome).

The final option is that we make bonkers backwards incompatible changes to storageSettings, but that's just something we have to control for by proper engineering considerations in future changes. I don't see that much more that can be added here to safeguard for that outcome.

Reply:
Gotcha, thanks!